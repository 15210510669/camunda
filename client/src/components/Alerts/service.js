import {get, post, del, put} from 'request';

export async function loadAlerts() {
  const response = await get('/api/alert');

  return await response.json();
}

export async function loadReports() {
  const response = await get('/api/report');

  return await response.json();
}

export async function saveNewAlert(alert) {
  const response = await post('/api/alert', alert);

  const json = await response.json();

  return json.id;
}

export async function updateAlert(id, alert) {
  return await put('/api/alert/' + id, alert);
}

export async function deleteAlert(id) {
  return await del('api/alert/' + id);
}

