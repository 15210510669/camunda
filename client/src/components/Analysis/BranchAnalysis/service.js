/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {post} from 'request';

export async function loadFrequencyData(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter
) {
  const response = await post(
    'api/report/evaluate',
    createFlowNodeFrequencyReport(
      processDefinitionKey,
      processDefinitionVersions,
      tenantIds,
      filter
    )
  );

  return await response.json();
}

function createFlowNodeFrequencyReport(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter
) {
  return {
    combined: false,
    reportType: 'process',
    data: {
      processDefinitionKey,
      processDefinitionVersions,
      tenantIds,
      filter,
      view: {
        entity: 'flowNode',
        property: 'frequency',
      },
      groupBy: {
        type: 'flowNodes',
        unit: null,
      },
      visualization: 'heat',
    },
  };
}

export async function loadCorrelationData(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter,
  gateway,
  end
) {
  const response = await post('api/analysis/correlation', {
    processDefinitionKey,
    processDefinitionVersions,
    tenantIds,
    filter,
    gateway,
    end,
  });

  return await response.json();
}
