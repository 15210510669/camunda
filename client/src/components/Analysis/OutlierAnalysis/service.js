/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get} from 'request';

export async function loadNodesOutliers(config) {
  const response = await get('api/analysis/flowNodeOutliers', config);
  return await response.json();
}

export {loadDurationData, loadCommonOutliersVariables, getInstancesDownloadUrl} from './service.ts';
