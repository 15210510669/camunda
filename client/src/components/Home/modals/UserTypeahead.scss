.UserTypeahead .DropdownOption .subTexts {
  display: block;
  font-size: 14px;
  color: grey;

  .subText {
    &:after {
      content: ' - ';
    }

    &:last-of-type:after {
      content: '';
    }
  }
}
