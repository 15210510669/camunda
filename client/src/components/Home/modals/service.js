/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {get} from 'request';
import {t} from 'translation';

export async function searchIdentities(terms) {
  const response = await get(`api/identity/search`, {terms});

  return await response.json();
}

export async function getDefinitionTenants(defintionKey, defintionType) {
  const response = await get(`api/definition/${defintionType}/${defintionKey}`);

  return await response.json();
}

export async function getDefinitionsWithTenants() {
  const response = await get('api/definition');

  return await response.json();
}

export async function getTenantsWithDefinitions() {
  const response = await get('api/definition/_groupByTenant');

  return await response.json();
}

export function formatDefintionName({name, type}) {
  const typeLabel = type === 'process' ? t('home.sources.process') : t('home.sources.decision');

  return name + ' (' + typeLabel + ')';
}
