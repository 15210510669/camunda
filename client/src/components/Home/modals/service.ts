/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get} from 'request';
import {Definition, Tenant} from 'types';

export type TenantWithDefinitions = Tenant & {definitions: Definition};

export type DefinitionWithTenants = Omit<Definition, 'key'> & {tenants: Tenant[]; key: string};

export async function getDefinitionsWithTenants(): Promise<DefinitionWithTenants[]> {
  const response = await get('api/definition');

  return await response.json();
}

export async function getTenantsWithDefinitions(): Promise<TenantWithDefinitions[]> {
  const response = await get('api/definition/_groupByTenant');

  return await response.json();
}
