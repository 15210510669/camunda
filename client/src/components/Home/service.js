/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {get, post, put, del} from 'request';

export async function loadEntities() {
  const response = await get('api/entities');
  return await response.json();
}

export async function addUser(collection, id, type, role) {
  return await post(`api/collection/${collection}/role`, {identity: {id, type}, role});
}

export async function editUser(collection, id, role) {
  return await put(`api/collection/${collection}/role/${id}`, {role});
}

export async function removeUser(collection, id) {
  return await del(`api/collection/${collection}/role/${id}`);
}

export async function loadAlerts(collection) {
  const response = await get(`api/collection/${collection}/alerts`);
  return await response.json();
}

export async function addAlert(alert) {
  return await post(`api/alert`, alert);
}

export async function editAlert(id, alert) {
  return await put(`api/alert/${id}`, alert);
}

export async function removeAlert(id) {
  return await del(`api/alert/${id}`);
}

export async function copyEntity(type, id, name, collectionId) {
  const query = {name};

  if (collectionId || collectionId === null) {
    query.collectionId = collectionId;
  }

  const response = await post(`api/${type}/${id}/copy`, undefined, {query});
  const json = await response.json();

  return json.id;
}

export async function searchIdentities(terms) {
  const response = await get(`api/identity/search`, {terms});

  return await response.json();
}
