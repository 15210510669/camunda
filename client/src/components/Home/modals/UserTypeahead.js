/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import debounce from 'debounce';
import {searchIdentities} from './service';
import {Typeahead} from 'components';
import {t} from 'translation';
import './UserTypeahead.scss';

export default class UserTypeahead extends React.Component {
  state = {
    loading: false,
    hasMore: false,
    initialDataLoaded: false,
    empty: false,
    identities: []
  };

  componentDidMount() {
    this.loadNewValues('');
  }

  loadNewValues = query => {
    if (this.state.initialDataLoaded && !query) {
      return this.cancelPendingSearch();
    }
    this.setState({loading: true});
    this.search(query);
  };

  search = debounce(async query => {
    const {total, result} = await searchIdentities(query);
    this.setState({
      identities: result,
      loading: false,
      hasMore: total > result.length,
      empty: result.length === 0,
      initialDataLoaded: !query
    });
  }, 800);

  cancelPendingSearch = () => {
    this.search.clear();
    this.setState({loading: false});
  };

  handleClose = () => {
    const {empty, loading} = this.state;
    // prevents unnecessary requests
    if (loading) {
      this.cancelPendingSearch();
    } else if (empty) {
      // prevents disabling the typeahead if closed empty
      this.setState({loading: true});
    }
  };

  selectIdentity = id => {
    const selectedIdentity = this.state.identities.find(identity => identity.id === id);
    if (selectedIdentity) {
      this.props.onChange(selectedIdentity);
    } else {
      this.props.onChange({id});
    }
  };

  render() {
    const {loading, hasMore, identities} = this.state;
    return (
      <Typeahead
        className="UserTypeahead"
        onSearch={this.loadNewValues}
        loading={loading}
        hasMore={!loading && hasMore}
        onClose={this.handleClose}
        onOpen={() => this.loadNewValues('')}
        placeholder={t('common.collection.addUserModal.searchPlaceholder')}
        onChange={this.selectIdentity}
        async
        typedOption
      >
        {identities.map(identity => {
          const {text, tag, subTexts} = formatTypeaheadOption(identity);
          return (
            <Typeahead.Option key={identity.id} value={identity.id} label={text}>
              <Typeahead.Highlight>{text}</Typeahead.Highlight>
              {tag}
              {subTexts && (
                <span className="subTexts">
                  {subTexts
                    .filter(subText => subText)
                    .map((subText, i) => (
                      <span className="subText" key={i}>
                        <Typeahead.Highlight matchFromStart>{subText}</Typeahead.Highlight>
                      </span>
                    ))}
                </span>
              )}
            </Typeahead.Option>
          );
        })}
      </Typeahead>
    );
  }
}

function formatTypeaheadOption({name, email, id, type}) {
  const subTexts = [];
  if (name) {
    subTexts.push(email);
  }

  if (name || email) {
    subTexts.push(id);
  }

  return {
    text: name || email || id,
    tag: type === 'group' && ' (User Group)',
    subTexts
  };
}
