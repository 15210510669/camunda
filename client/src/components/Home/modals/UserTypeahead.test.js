/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import UserTypeahead from './UserTypeahead';
import {searchIdentities} from './service';

jest.mock('debounce', () =>
  jest.fn(fn => {
    fn.clear = jest.fn();
    return fn;
  })
);

jest.mock('./service', () => ({
  searchIdentities: jest.fn().mockReturnValue({result: [], total: 50})
}));

const props = {
  onChange: jest.fn()
};

it('should render a Typeahead', () => {
  const node = shallow(<UserTypeahead {...props} />);

  expect(node.find('Typeahead')).toExist();
});

it('should load initial data when component is mounted', () => {
  shallow(<UserTypeahead />);

  expect(searchIdentities).toHaveBeenCalledWith('');
});

it('should enable loading while loading data and enable hasMore if there are more data available', async () => {
  const node = shallow(<UserTypeahead {...props} />);

  expect(node.find('Typeahead').prop('loading')).toBe(true);
  await node.update();
  expect(node.find('Typeahead').prop('loading')).toBe(false);
  expect(node.find('Typeahead').prop('hasMore')).toBe(true);
});

it('should enable loading if typeahead is closed while empty', async () => {
  const node = shallow(<UserTypeahead />);

  node.find('Typeahead').prop('onSearch')('test');
  await node.update();
  node.find('Typeahead').prop('onClose')();
  expect(node.find('Typeahead').prop('loading')).toBe(true);
});

it('should format user list information correctly', () => {
  const node = shallow(<UserTypeahead {...props} />);

  node.setState({
    identities: [
      {id: 'testUser'},
      {id: 'user2', email: 'testUser@test.com'},
      {id: 'groupId', name: 'groupName', email: 'group@test.com', type: 'group'}
    ]
  });

  expect(node).toMatchSnapshot();
});
