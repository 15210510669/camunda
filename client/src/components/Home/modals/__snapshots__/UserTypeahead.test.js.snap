// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should format user list information correctly 1`] = `
<Typeahead
  async={true}
  className="UserTypeahead"
  hasMore={false}
  loading={true}
  onChange={[Function]}
  onClose={[Function]}
  onOpen={[Function]}
  onSearch={[Function]}
  placeholder="Search for name, email or ID ..."
>
  <ForwardRef(DropdownOption)
    key="testUser"
    label="testUser"
    value="testUser"
  >
    testUser
    <span
      className="subTexts"
    />
  </ForwardRef(DropdownOption)>
  <ForwardRef(DropdownOption)
    key="user2"
    label="testUser@test.com"
    value="user2"
  >
    testUser@test.com
    <span
      className="subTexts"
    >
      <span
        className="subText"
        key="0"
      >
        user2
      </span>
    </span>
  </ForwardRef(DropdownOption)>
  <ForwardRef(DropdownOption)
    key="groupId"
    label="groupName"
    value="groupId"
  >
    groupName
     (User Group)
    <span
      className="subTexts"
    >
      <span
        className="subText"
        key="0"
      >
        group@test.com
      </span>
      <span
        className="subText"
        key="1"
      >
        groupId
      </span>
    </span>
  </ForwardRef(DropdownOption)>
</Typeahead>
`;
